core/src/test/scala/org/apache/spark/DataPropertyAccumulatorSuite.scala

```diff
@@ -0,0 +1,361 @@
+  test("single partition") {
+    sc = new SparkContext("local[2]", "test")
+    val acc : Accumulator[Int] = sc.accumulator(0, dataProperty = true)
+
+    val a = sc.parallelize(1 to 20, 1)
+    val b = a.map{x => acc += x; x}
+    b.cache()
+    b.count()
+    acc.value should be (210)
+  }
+
+  test("adding only the first element per partition should work even if partition is empty") {
+    sc = new SparkContext("local[2]", "test")
+    val acc: Accumulator[Int] = sc.accumulator(0, dataProperty = true)
+    val a = sc.parallelize(1 to 20, 30)
+    val b = a.mapPartitions{itr =>
+      acc += 1
+      itr
+    }
+    b.count()
+    acc.value should be (30)
+  }
+
+  test("shuffled (combineByKey)") {
+    sc = new SparkContext("local[2]", "test")
+    val a = sc.parallelize(1 to 40, 5)
+    val buckets = 4
+    val b = a.map{x => ((x % buckets), x)}
+    val inputs = List(b, b.repartition(10), b.partitionBy(new HashPartitioner(5))).map(_.cache())
+    val mapSideCombines = List(true, false)
+    inputs.foreach { input =>
+      mapSideCombines.foreach { mapSideCombine =>
+        val accs = (1 to 4).map(x => sc.accumulator(0, dataProperty = true)).toList
+        val raccs = (1 to 4).map(x => sc.accumulator(0, dataProperty = false)).toList
+        val List(acc, acc1, acc2, acc3) = accs
+        val List(racc, racc1, racc2, racc3) = raccs
+        val c = input.combineByKey(
+          (x: Int) => {acc1 += 1; acc += 1; racc1 += 1; racc += 1; x},
+          {(a: Int, b: Int) => acc2 += 1; acc += 1; racc2 += 1; racc += 1; (a + b)},
+          {(a: Int, b: Int) => acc3 += 1; acc += 1; racc3 += 1; racc += 1; (a + b)},
+          new HashPartitioner(10),
+          mapSideCombine)
+        val d = input.combineByKey(
+          (x: Int) => {acc1 += 1; acc += 1; x},
+          {(a: Int, b: Int) => acc2 += 1; acc += 1; (a + b)},
+          {(a: Int, b: Int) => acc3 += 1; acc += 1; (a + b)},
+          new HashPartitioner(2),
+          mapSideCombine)
+        val e = d.map{x => acc += 1; x}
+        c.count()
+        // If our partitioner is known then we should only create
+        // one combiner for each key value. Otherwise we should
+        // create at least that many combiners.
+        if (input.partitioner.isDefined) {
+          acc1.value should be (buckets)
+        } else {
+          acc1.value should be >= (buckets)
+        }
+        if (input.partitioner.isDefined) {
+          acc2.value should be > (0)
+        } else if (mapSideCombine) {
+          acc3.value should be > (0)
+        } else {
+          acc2.value should be > (0)
+          acc3.value should be (0)
+        }
+        acc.value should be (acc1.value + acc2.value + acc3.value)
+        val oldValues = accs.map(_.value)
+        // For one action the data property accumulators and regular should have the same value.
+        accs.map(_.value) should be (raccs.map(_.value))
+        c.count()
+        accs.map(_.value) should be (oldValues)
```

Comment:
This is interesting. I tried to do the same behavior with different approach. I check RDD ID ahd Shuffle ID at DAGScheduler. Accumulators are not allowed to be updated on the same RDD ID and Shuffle ID. This simpler approach can achieve part function of this DataProperty accumulators.

However, it can not handle some cases. For example, it will not pass all checks in the test `shuffled (combineByKey)`.  To put accumulators in shuffling aggregator looks like a very rare use case. As the aggregation will perform when ShuffleRDD is going to read shuffled output. It will not generate a new Shuffle ID and that accumulator updates in shuffling aggregator will be counted again.